### ISSUE TITLE
Accuracy on Validation Set

### ISSUE THREAD

**Issue by SparkJiao** [SKTBrain/SUMBT#3]
Hi, sorry to bother you.
Because the training time of 300 epochs is much long, I am trying a training procedure with `learning_rate=1e-4, num_train_epochs=100` and to save memory so that the model could run on a single GPU with 12 GB memory, I set `train_batch_size=3` and `eval_batch_size=4`.
Currently, at the 45th epoch, the joint accuracy on validation set has reached 53.96%.  And the model is still on warm up stage. Is it normal? Or there are some problems?  
![image](https://user-images.githubusercontent.com/16469472/63071016-01343b80-bf50-11e9-9c58-567a071b1012.png)
![image](https://user-images.githubusercontent.com/16469472/63071027-101aee00-bf50-11e9-8ee5-6a155e2314ec.png)
![image](https://user-images.githubusercontent.com/16469472/63071042-288b0880-bf50-11e9-8234-4c54cb2f9290.png)
Besides, the dataset I used is MultiWOZ2.1.
Thank you very much.

**Comment by SparkJiao** [comment]
I have found the bug of learning rate. The`num_train_steps` should be calculated with the total number of train features, not train examples, since a single step equals `batch_size`, not `batch_size * max_dialog_turns`.

**Comment by SparkJiao** [comment]
I'm sorry and I have figure out why the accuracy on validation set and test set is higher than the performance reported in the paper. I think it's because all undefined values in ontology are set `none` but indeed these predictions will be wrong all the time.
So would you mind relasing the script to evaluate the predictions? Or could you please tell me how to get the results on your paper?
Thank you very much!

**Comment by hwaranlee** [comment]
Hi @SparkJiao,
Thank you for sharing the experiment results on MultiWOZ2.1. So your experiment result is higher than our results reported in the paper, right?
1) Even though the model is on the warm-up stage, the train accuracy can reach the ceil.
2) The new MultiWOZ 2.1 corpus maybe easier than MultiWOZ2.0 which we expereimented on. Or your experiment configure is more effective than ones we did.
3) For the `num_trainn_steps` issue, I don't understand what you mean.
4) You can get the evaluation result by the command with the option `--do_eval`, in which we also achieve the results.

**Comment by SparkJiao** [comment]
Hi, @hwaranlee 
First of all, I have modified the way to calculate the variable `num_train_steps`.
Here is the way in your code:
```
num_train_steps = int(len(train_examples) / args.train_batch_size / args.gradient_accumulation_steps * args.num_train_epochs)
...
t_total = num_train_steps
```
And while updating learning rate:
```
lr_this_step = args.learning_rate * warmup_linear(global_step / t_total, args.warmup_proportion)
```
So a single step should be a group of batch, which has `batch_size` train features, not train examles. A feature contains `max_turn` train examples. So the total step is much larger than it should be. As a result, the learning rate is linear increasing all the time and you know the model is difficult to get convergence with a big learning rate.

Indeed, I change it like this:

```
 num_train_features = all_input_ids.size(0)
 num_train_steps = int(num_train_features / args.train_batch_size / args.gradient_accumulation_steps * args.num_train_epochs)
```

So I have few experiments. After fix the way calculating `num_train_steps`, I re-run it on **Multiwoz2.0** with following parameters:
```
learning_rate=1e-4, num_train_epochs=100, train_batch_size=3
```
and I got a joint accuracy of 50.136%. Indeed the model stop training at around 45 epochs.

Then I fix the undefined value, in which for undefined values, I simply mark the predictions as wrong(because the prediction are always pre-definded in ontology). And using the same parameters, I got a joint accuracy of 48.386%. Both on test set.
So after fixing the error the performance actually drop. I'm still running your code on multiwoz2.0 to reproduce your results. If the results is the same as that on your paper, then the actual results should be lower but after fix the learning rate bug it will be much better. 
BTW, if you decide to fix the bug and re-run it I think your will get a result rapidly than me.
Thanks for your reply.

**Comment by hwaranlee** [comment]
Thank you for your comment!
For the learning rate and num_train_steps, yes you are right! I fix the bug and pushed it.
With the same parameters, I got 48.80% joint accuracy on the test set.

> Then I fix the undefined value, in which for undefined values, I simply mark the predictions as wrong(because the prediction are always pre-definded in ontology).

Do you mean that slot values having 'none' labels since they are not undefined in the ontology? For this, we follow the given pre-defined ontology. Since the ontology is incomplete and does not cover all slot-values, so the way of prediction is debatable. However, we assume the ontology is complete, even though it is strong assumption, and we considered the lables of undefined slot-values as 'none' class.

**Comment by SparkJiao** [comment]
Thanks a lot, I have known your setting.

Apperaciate to your contribution!
